### DISCUSSION TITLE
Replaced all deflated images (of a certain type) with jpeg equivalent

### DISCUSSION THREAD

**Discussion by HinTak** [pymupdf/PyMuPDF#3464] [Looking for help] [answered]
I feel this could be an FAQ but couldn't find a relevant switch/example. What is the mutool convert / pymupdf equivalent of wanting these with ghostscript's pdfwrite, or combinations of?
-dDownsampleColorImages=true				
-dDownsampleGrayImages=true
-dDownsampleMonoImages=true

I.e. basically replacing zlib deflated images with their jpeg / ccitt tiff equivalent.

**Comment by HinTak** [comment]
Actually downsampling isn't exactly what I want - I want to convert all images from zlib deflate to dctdecode/ccittdecode, at its original resolution. This seems to be a useful /frequent request?

**Comment by JorjMcKie** [comment]
We currently have a - possible meagre - version for that: `page.replace_image(xref, <new image spec>)` re-inserts an arbitrary new image under the xref of the existing one.
So you could extract a given image, massage it some desired way and insert it again as if it would be new.

**Comment by HinTak** [comment, marked as answer]
Thanks for the tips . Here is what i come up with now: https://github.com/HinTak/pymupdf-jbig2-extract/blob/main/lossy-convert.py

It is mostly doing want I wanted - just convert RGB flatedecoded images to jpeg. I have two questions though:

- going via `PIL.Image.save` vs `fitz.Pixmap.save` results in much better compression, by default even without the `optimized=True` PIL key (you can roll back a commit or two before to see the diff - I started off with PIL actually, then wanted to remove that dependency and found the size result to be worse - the public repo has simplified/reverted history compared to my private repo). In one file I use for such test, original is 400MB. Going via PIL gives a 75MB output (and down to 72MB with `optimized=True`), while via `fitz.Pixmap.save` gives a 200MB pdf. I am quite surprised by this, since it should be compressing a large raw bitmap in both cases, and should just be libjpeg's default, and may even be identical (except for padding at image edges). Why is `Pixmap.save` so much worse? Or, what settings are not default? I seem to remember Adobe has different ideas about jpeg quality setting from libjpeg's.

- the other thing I like to ask is what's mupdf's equivalent for not updating.the 2nd field in the ID with a random number? `qpdf` offers a `--deterministic-ID` . This should probably be a `doc.save` option.

**Comment by HinTak** [comment]
Argh, I have my answer to the 2nd question: `no_new_id=True`.

**Comment by HinTak** [comment]
Argh, have the answer to my first question also, after extracting the corresponding objects. Upstream mupdf set progressive=True, set dpi=96, disable chroma subsampling, and set quality to 95. Doing all these in PIL I cab get bitwise identical output. Found all of them in upstream mupdf code too, except the last one: upstream default to 90. 95 is a pymupdf setting?

**Comment by HinTak** [comment]
Yes, 95 is a pymupdf setting: https://github.com/pymupdf/PyMuPDF/blob/d46413390cbdbae7c3ea6a9030a9f3e71ecc42ff/docs/pixmap.rst?plain=1#L326

This concludes everything I want to know now. I'll add these as comments, but keep the PIL code.
